### PR TITLE
Factory

### PR DESCRIPTION
**Pull request by marondeau-mila**
# Start of factorization

The goal is to separate the creation of object from their use. This would reduce the boilerplate than need to change for the project, and simplify configuration.
Since the optimizer and scheduler are complex, and their creation must follow strict rules under Lightning, they were factorized first.

## Boilerplate
With factories, we can trivially change the algorithm, and associated hyper-parameters without having to change the modules: the factory knows the details, and the module knows how to call the factory. Changing the algorithm only requires changing things once: where the configuration is loaded.
Even better: checkpoints will be trivial to load, since the factory is saved with the model.

## Configuration
This change removes the need for configuration deep in the modules. This means that using OmniConf, Hydra, a complex set of command line argument, or some HTTP GET based abomination can be done without having to change the Module.
This also means that the Module does not need to know about the hyper-parameters. We effectively move them around inside the factory.

## Limitations and Future Work
In general, using factories will simplify the inner part of the project: the Module and training loop. We still need to manage configuration, to create the factory. This is a current challenge with the template, which won't suddenly disappear with factories. Factories would help addressing that challenge, by decoupling the configuration from the Module.

The factory approach would also be useful for the loss and model. This PR includes some preliminary refactorization in that direction, but the main goal is to get feedback on the structure and use cases.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[flake8]
+max-line-length=100
```
Comment (hbertrand) [inline .flake8:2]: 100 is very small. Can we increase it?

File: amlrt_project/models/factory.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+class PlateauFactory(SchedulerFactory):
+    """Reduce the learning rate when `metric` is no longer improving."""
+    metric: str
+    """Metric to use, must be logged with Lightning."""
+    mode: str = "min"
+    """Minimize or maximize."""
+    factor: float = 0.1
+    """Multiply the learning rate by `factor`."""
+    patience: int = 10
+    """Wait `patience` epoch before reducing the learning rate."""
```
Comment (hbertrand) [inline amlrt_project/models/factory.py:61]: Is this how we are supposed to document fields in dataclasses? I find it hard to read.

File: amlrt_project/models/model_loader.py [hunk 3]
```diff
@@ -1,11 +1,72 @@
...
+        raise ValueError(f'Optimizer {algo} not supported')
+    else:
+        algo: Type[OptimFactory] = OPTS[algo]
+
+    return algo(**args)
+
+
+def parse_sched_hp(
+    hyper_params: Optional[Union[str, Dict[str, Any]]]
+) -> Optional[SchedulerFactory]:
```
Comment (hbertrand) [inline amlrt_project/models/model_loader.py:48]: ugliness caused by the small line length?

File: amlrt_project/models/my_model.py [hunk 4]
```diff
@@ -62,22 +80,25 @@ class SimpleMLP(BaseModel):  # pragma: no cover
 
     Inherits from the given framework's model class. This is a simple MLP model.
     """
-    def __init__(self, hyper_params: typing.Dict[typing.AnyStr, typing.Any]):
+    def __init__(
+        self,
+        optim_fact: OptimFactory,
+        sched_fact: SchedulerFactory,
```
Comment (hbertrand) [inline amlrt_project/models/my_model.py:86]: should go into kwargs and just forwarded to parent?

**Review by hbertrand — COMMENTED** [review]
I like the factory design, it should make things quite a bit simpler over a long project.

File: .flake8 (hunk 1 above)
Comment (marondeau-mila) [reply .flake8:2]: Yes, that's just to match the value used in the CI, that way my editor warns me when I save the file.

File: amlrt_project/models/model_loader.py (hunk 3 above)
Comment (marondeau-mila) [reply amlrt_project/models/model_loader.py:48]: Yep.

File: amlrt_project/models/factory.py (hunk 2 above)
Comment (marondeau-mila) [reply amlrt_project/models/factory.py:61]: We could add lines and mess with the formatting, and we could also put everything in the main docstring.
There's no standard way of documenting fields. The advantage of that one is that editors and sphinx understand it. It seems like they don't show up when looking at the constructor.
The other advantage is that the documentation is next to the field, so it's easy to see what is undocumented, and to keep it up to date.

File: amlrt_project/models/my_model.py (hunk 4 above)
Comment (marondeau-mila) [reply amlrt_project/models/my_model.py:86]: No, they are needed, and I want them explicit. That way, we get an useful error message if they are missing.

Comment (hbertrand) [reply amlrt_project/models/my_model.py:86]: If they are missing you'll just get the error message from the parent. But I agree that it won't matter anyway in the fully factorized version.

**Review by hbertrand — APPROVED** [review]

File: amlrt_project/models/factory.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    r"""Length of the linear warmup."""
+    eps: float = field(init=False, default=1e-16)
+    r"""Safety value: `gamma` must be larger than this."""
+
+    def __post_init__(self):
+        """Finish initialization."""
+        self.gamma = max(min(self.gamma, 1.0), self.eps)
+        # Clip gamma to something that make sense, just in case.
+        self.warmup = max(self.warmup, 0)
+        # Same for warmup.
```
Comment (mirkobronzi) [inline amlrt_project/models/factory.py:97]: @marondeau-mila : what is the convention for the comment position?
In this example, the comment (97) follows the line (96). (but also 95 and 94)
Below, in `amlrt_project/models/my_model.py`, the comment (96) is before the line (97).

Is there a reason for this? I usually see the second option (comment just before the line) as the main choice.

Comment (mirkobronzi) [reply amlrt_project/models/factory.py:97]: Discussed with @marondeau-mila - the point is that there is no standard w.r.t. this, and also that doc extractor tool (e.g., Sphynx) will behave differently.
@hbertrand : any comment on this?

Comment (hbertrand) [reply amlrt_project/models/factory.py:97]: I don't care much which one we choose as long as we are consistent within the repo.

Comment (mirkobronzi) [reply amlrt_project/models/factory.py:97]: it is not consistent already (first message).

Comment (mirkobronzi) [reply amlrt_project/models/factory.py:96]: fixed - we use the standard where a comment is before the related line of code

**Review by mirkobronzi — APPROVED** [review]

**Review by hbertrand — APPROVED** [review]
